### PR TITLE
Add backend implementation for bot suggestions

### PR DESCRIPTION
**Pull request by siggisim**
Only enabled if an OpenAI api key is provided in your config.yaml

---

**Version bump**: Patch

### CODE REVIEW

**Review by sluongng — APPROVED** [review]
image

This is so so so cool.
I am HYPED!

File: enterprise/server/suggestion/suggestion.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+	"github.com/buildbuddy-io/buildbuddy/server/eventlog"
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+
+	elpb "github.com/buildbuddy-io/buildbuddy/proto/eventlog"
+	supb "github.com/buildbuddy-io/buildbuddy/proto/suggestion"
+)
+
+var apiKey = flag.String("openai.api_key", "", "OpenAI API key")
+var model = flag.String("openai.model", "gpt-3.5-turbo", "OpenAI model name to use, if not set 'gpt-3.5-turbo' is used.")
```
Comment (tylerwilliams) [inline enterprise/server/suggestion/suggestion.go:22]: nit: you don't need the default in the help text, the flags lib will show it in the --help output like this:

```
  -telemetry_endpoint string
    	The telemetry endpoint to use (default "grpcs://t.buildbuddy.io:443")
  -telemetry_interval duration
    	How often telemetry data will be reported (default 24h0m0s)
```

File: enterprise/server/suggestion/suggestion.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+const (
+	minLines = 1000
+	maxChars = 8000
```
Comment (tylerwilliams) [inline enterprise/server/suggestion/suggestion.go:26]: nit: could use a comment here about what these are and how to set them

File: enterprise/server/suggestion/suggestion.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+	res := &supb.GetSuggestionResponse{
+		Suggestion: []string{completionResponse.Choices[0].Message.Content},
+	}
+
+	return res, nil
+}
+
+// TODO(siggisim): Pull this into its own backend if we want to use this in other places.
+func getCompletions(data *completionRequest) (*completionResponse, error) {
+	url := "https://api.openai.com/v1/chat/completions"
```
Comment (tylerwilliams) [inline enterprise/server/suggestion/suggestion.go:96]: nit: can be const

File: enterprise/server/suggestion/suggestion.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+	}
+	defer postResp.Body.Close()
+
+	body, err := io.ReadAll(postResp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if postResp.StatusCode != http.StatusOK {
+		log.Printf("%+v %+v", postResp.StatusCode, string(body))
```
Comment (tylerwilliams) [inline enterprise/server/suggestion/suggestion.go:123]: Maybe log.Debugf?

File: enterprise/server/suggestion/suggestion.go (hunk 3 above)
Comment (siggisim) [reply enterprise/server/suggestion/suggestion.go:96]: Addressed in: https://github.com/buildbuddy-io/buildbuddy/pull/3583/commits/2f05336a0f6f4d071e240592e93e459888a62324

File: enterprise/server/suggestion/suggestion.go (hunk 4 above)
Comment (siggisim) [reply enterprise/server/suggestion/suggestion.go:123]: Addressed in: https://github.com/buildbuddy-io/buildbuddy/pull/3583/commits/2f05336a0f6f4d071e240592e93e459888a62324

File: enterprise/server/suggestion/suggestion.go (hunk 2 above)
Comment (siggisim) [reply enterprise/server/suggestion/suggestion.go:26]: Addressed in: https://github.com/buildbuddy-io/buildbuddy/pull/3583/commits/2f05336a0f6f4d071e240592e93e459888a62324

File: enterprise/server/suggestion/suggestion.go (hunk 1 above)
Comment (siggisim) [reply enterprise/server/suggestion/suggestion.go:22]: That's awesome, addressed in: https://github.com/buildbuddy-io/buildbuddy/pull/3583/commits/2f05336a0f6f4d071e240592e93e459888a62324

**Pull request merged**